### PR TITLE
[tools] add mixed precision training

### PR DESCRIPTION
**Pull request by whiteshirt0429**

### CODE REVIEW

**Comment by Gqwert123** [comment]
老哥，你的模型forward那里是不是也要包一下

**Comment by whiteshirt0429** [comment]
> 老哥，你的模型forward那里是不是也要包一下

哪里 能帮忙指出来吗 文件多少行

**Comment by fanlu** [comment]
The latest result trained by 4 V100 in one node
|CER|attention_rescoring|attention|ctc_greedy|ctc_prefix_beam|
|--|--|--|--|--|
||5.01|5.40|5.53|5.52|

**Comment by whiteshirt0429** [comment]
| decoding mode          | CER  |
|------------------------|------|
| attention decoder      | 5.81 |
| ctc greedy search      | 6.14 |
| ctc prefix beam search | 6.14 |
| attention rescoring    | 5.48 |

**Comment by whiteshirt0429** [comment]
transformer model with 46 batchszie, lr : 0.004 ,4gpu
| decoding mode          | CER  |
|------------------------|------|
| attention decoder      | 5.81 |
| ctc greedy search      | 6.14 |
| ctc prefix beam search | 6.14 |
| attention rescoring    | 5.48 |

**Comment by whiteshirt0429** [comment]
second run transformer model with batchsize: 52, lr 0.002, the error is less than 0.05
| decoding mode          | CER  |
|------------------------|------|
| attention decoder      | 5.75 |
| ctc greedy search      | 5.93 |
| ctc prefix beam search | 5.93 |
| attention rescoring    | 5.35 |

File: wenet/utils/executor.py [hunk 1]
```diff
@@ -22,8 +23,12 @@ def train(self, model, optimizer, scheduler, data_loader, device, writer,
         rank = args.get('rank', 0)
         accum_grad = args.get('accum_grad', 1)
         is_distributed = args.get('is_distributed', True)
+        use_amp = args.get('use_amp', False)
         logging.info('using accumulate grad, new batch size is {} times'
                      'larger than before'.format(accum_grad))
+        print(use_amp)
```
Comment (robin1001) [inline wenet/utils/executor.py:29]: remove this

**Comment by whiteshirt0429** [comment]
average_num 20 results
| decoding mode          | CER  |
|------------------------|------|
| attention decoder      | 5.73 |
| ctc greedy search      | 5.92 |
| ctc prefix beam search | 5.92 |
| attention rescoring    | 5.31 |

File: wenet/utils/executor.py [hunk 2]
```diff
@@ -39,31 +43,50 @@ def train(self, model, optimizer, scheduler, data_loader, device, writer,
...
                 if rank == 0 and writer is not None:
                     writer.add_scalar('train_loss', loss, self.step)
-                grad_norm = clip_grad_norm_(model.parameters(), clip)
-                if torch.isfinite(grad_norm):
-                    optimizer.step()
+                # Use mixed precision training
+                if use_amp:
+                    scaler.unscale_(optimizer)
+                    grad_norm = clip_grad_norm_(model.parameters(), clip)
+
```
Comment (robin1001) [inline wenet/utils/executor.py:73]: delete the blank line

**Review by robin1001 — APPROVED** [review]

**Pull request merged**